File: packages/engine/src/ai/heuristic.ts
/** AI.md §2 — line-decomposable heuristic with integer weights. */
import { column, nibble } from './bitboard.ts';

export interface HeuristicWeights {
  lost: number;
  empty: number;
  merges: number;
  mono: number;
  sum: number;
  smooth: number;
  stable: number;
  corner: number;
}

export const HEURISTIC_V1: Readonly<HeuristicWeights> = Object.freeze({
  lost: 200000,
  empty: 270,
  merges: 700,
  mono: 47,
  sum: 11,
  smooth: 0,
  stable: 0,
  corner: 0,
});

export interface LineFeatures {
  empty: number;
  merges: number;
  mono: number;
  sum: number;
  smooth: number;
  stable: number;
}

export function lineFeatures(v: number): LineFeatures {
  const r = [v & 0xf, (v >> 4) & 0xf, (v >> 8) & 0xf, (v >> 12) & 0xf];
  let empty = 0;
  let merges = 0;
  let prev = 0;
  let counter = 0;
  let sum = 0;
  for (const rank of r) {
    sum += rank * rank * rank;
    if (rank === 0) {
      empty++;
      continue;
    }
    if (prev === rank) counter++;
    else if (counter > 0) {
      merges += 1 + counter;
      counter = 0;
    }
    prev = rank;
  }
  if (counter > 0) merges += 1 + counter;

  let monoL = 0;
  let monoR = 0;
  let smooth = 0;
  for (let i = 1; i < 4; i++) {
    const a = r[i - 1] ** 4;
    const b = r[i] ** 4;
    if (r[i - 1] > r[i]) monoL += a - b;
    else monoR += b - a;
    if (r[i - 1] !== 0 && r[i] !== 0) smooth += Math.abs(r[i - 1] - r[i]);
  }
  const stable = empty === 0 && (monoL === 0 || monoR === 0) ? 1 : 0;
  return { empty, merges, mono: Math.min(monoL, monoR), sum, smooth, stable };
}

export function lineScore(f: LineFeatures, w: HeuristicWeights): number {
  return (
    w.lost +
    w.empty * f.empty +
    w.merges * f.merges -
    w.mono * f.mono -
    w.sum * f.sum -
    w.smooth * f.smooth +
    w.stable * f.stable
  );
}

export function validateWeights(w: HeuristicWeights): void {
  for (const [k, v] of Object.entries(w)) {
    if (!Number.isInteger(v)) throw new Error(`heuristic weight '${k}' must be an integer (got ${v})`);
  }
}

const tableCache = new Map<string, Float64Array>();

/** 65536-entry per-line score table for the given weights (memoised). */
export function lineTable(w: HeuristicWeights): Float64Array {
  const key = JSON.stringify([w.lost, w.empty, w.merges, w.mono, w.sum, w.smooth, w.stable, w.corner]);
  let t = tableCache.get(key);
  if (!t) {
    validateWeights(w);
    t = new Float64Array(65536);
    for (let v = 0; v < 65536; v++) t[v] = lineScore(lineFeatures(v), w);
    tableCache.set(key, t);
  }
  return t;
}

export function cornerTerm(lo: number, hi: number, cornerWeight: number): number {
  if (cornerWeight === 0) return 0;
  let max = 0;
  for (let i = 0; i < 16; i++) {
    const n = nibble(lo, hi, i);
    if (n > max) max = n;
  }
  const c0 = lo & 0xf;
  const c3 = (lo >>> 12) & 0xf;
  const c12 = (hi >>> 16) & 0xf;
  const c15 = hi >>> 28;
  return c0 === max || c3 === max || c12 === max || c15 === max ? cornerWeight * max : 0;
}

export function evaluate(lo: number, hi: number, table: Float64Array, cornerWeight: number): number {
  return (
    table[lo & 0xffff] +
    table[lo >>> 16] +
    table[hi & 0xffff] +
    table[hi >>> 16] +
    table[column(lo, hi, 0)] +
    table[column(lo, hi, 1)] +
    table[column(lo, hi, 2)] +
    table[column(lo, hi, 3)] +
    cornerTerm(lo, hi, cornerWeight)
  );
}

export interface HeuristicBreakdown extends LineFeatures {
  corner: number;
  total: number;
}

/** Human-readable decomposition of the evaluation, summed over all 8 lines. */
export function breakdown(lo: number, hi: number, w: HeuristicWeights): HeuristicBreakdown {
  const lines = [lo & 0xffff, lo >>> 16, hi & 0xffff, hi >>> 16, column(lo, hi, 0), column(lo, hi, 1), column(lo, hi, 2), column(lo, hi, 3)];
  const acc: LineFeatures = { empty: 0, merges: 0, mono: 0, sum: 0, smooth: 0, stable: 0 };
  for (const v of lines) {
    const f = lineFeatures(v);
    for (const k of Object.keys(acc) as (keyof LineFeatures)[]) acc[k] += f[k];
  }
  const corner = cornerTerm(lo, hi, w.corner);
  return { ...acc, corner, total: evaluate(lo, hi, lineTable(w), w.corner) };
}

File: engines/rust/src/ai/heuristic.rs
//! AI.md §2 — line-decomposable heuristic with integer weights.
use serde::Serialize;
use serde_json::{Map, Value};

use super::bitboard::{column, max_rank, nibble};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Weights {
    pub lost: i64,
    pub empty: i64,
    pub merges: i64,
    pub mono: i64,
    pub sum: i64,
    pub smooth: i64,
    pub stable: i64,
    pub corner: i64,
}

pub const HEURISTIC_V1: Weights = Weights {
    lost: 200000,
    empty: 270,
    merges: 700,
    mono: 47,
    sum: 11,
    smooth: 0,
    stable: 0,
    corner: 0,
};

impl Default for Weights {
    fn default() -> Weights {
        HEURISTIC_V1
    }
}

fn integer_weight(v: &Value, key: &str) -> Result<i64, String> {
    if let Some(i) = v.as_i64() {
        return Ok(i);
    }
    if let Some(f) = v.as_f64() {
        if f.is_finite() && f.fract() == 0.0 && f.abs() < 9.0e15 {
            return Ok(f as i64);
        }
    }
    Err(format!("heuristic weight '{}' must be an integer (got {})", key, v))
}

impl Weights {
    /// Overlay a (possibly partial) JSON weights object on top of `self`.
    pub fn merged_with(&self, v: &Value) -> Result<Weights, String> {
        let obj = match v.as_object() {
            Some(o) => o,
            None => return Err(String::from("weights must be an object")),
        };
        let mut w = *self;
        for (k, val) in obj.iter() {
            let x = integer_weight(val, k)?;
            match k.as_str() {
                "lost" => w.lost = x,
                "empty" => w.empty = x,
                "merges" => w.merges = x,
                "mono" => w.mono = x,
                "sum" => w.sum = x,
                "smooth" => w.smooth = x,
                "stable" => w.stable = x,
                "corner" => w.corner = x,
                _ => return Err(format!("unknown heuristic weight '{}'", k)),
            }
        }
        Ok(w)
    }

    pub fn from_json(v: &Value) -> Result<Weights, String> {
        HEURISTIC_V1.merged_with(v)
    }

    pub fn to_json(&self) -> Value {
        let mut m = Map::new();
        m.insert(String::from("lost"), Value::from(self.lost));
        m.insert(String::from("empty"), Value::from(self.empty));
        m.insert(String::from("merges"), Value::from(self.merges));
        m.insert(String::from("mono"), Value::from(self.mono));
        m.insert(String::from("sum"), Value::from(self.sum));
        m.insert(String::from("smooth"), Value::from(self.smooth));
        m.insert(String::from("stable"), Value::from(self.stable));
        m.insert(String::from("corner"), Value::from(self.corner));
        Value::Object(m)
    }
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize)]
pub struct LineFeatures {
    pub empty: i64,
    pub merges: i64,
    pub mono: i64,
    pub sum: i64,
    pub smooth: i64,
    pub stable: i64,
}

pub fn line_features(v: u16) -> LineFeatures {
    let r: [i64; 4] = [
        (v & 0xF) as i64,
        ((v >> 4) & 0xF) as i64,
        ((v >> 8) & 0xF) as i64,
        ((v >> 12) & 0xF) as i64,
    ];
    let mut empty: i64 = 0;
    let mut merges: i64 = 0;
    let mut prev: i64 = 0;
    let mut counter: i64 = 0;
    let mut sum: i64 = 0;
    for k in 0..4usize {
        let rank = r[k];
        sum += rank * rank * rank;
        if rank == 0 {
            empty += 1;
            continue;
        }
        if prev == rank {
            counter += 1;
        } else if counter > 0 {
            merges += 1 + counter;
            counter = 0;
        }
        prev = rank;
    }
    if counter > 0 {
        merges += 1 + counter;
    }

    let mut mono_l: i64 = 0;
    let mut mono_r: i64 = 0;
    let mut smooth: i64 = 0;
    for i in 1..4usize {
        let a = r[i - 1] * r[i - 1] * r[i - 1] * r[i - 1];
        let b = r[i] * r[i] * r[i] * r[i];
        if r[i - 1] > r[i] {
            mono_l += a - b;
        } else {
            mono_r += b - a;
        }
        if r[i - 1] != 0 && r[i] != 0 {
            smooth += (r[i - 1] - r[i]).abs();
        }
    }
    let stable: i64 = if empty == 0 && (mono_l == 0 || mono_r == 0) { 1 } else { 0 };
    LineFeatures {
        empty,
        merges,
        mono: if mono_l < mono_r { mono_l } else { mono_r },
        sum,
        smooth,
        stable,
    }
}

/// Exact integer line score (all intermediates are far below 2^53).
pub fn line_score(f: &LineFeatures, w: &Weights) -> f64 {
    let s: i64 = w.lost + w.empty * f.empty + w.merges * f.merges
        - w.mono * f.mono
        - w.sum * f.sum
        - w.smooth * f.smooth
        + w.stable * f.stable;
    s as f64
}

/// 65536-entry per-line score table for the given weights.
pub fn line_table(w: &Weights) -> Vec<f64> {
    let mut t: Vec<f64> = vec![0.0f64; 65536];
    for v in 0..65536usize {
        t[v] = line_score(&line_features(v as u16), w);
    }
    t
}

pub fn corner_term(b: u64, corner_weight: i64) -> f64 {
    if corner_weight == 0 {
        return 0.0;
    }
    let max = max_rank(b) as u8;
    let c0 = nibble(b, 0);
    let c3 = nibble(b, 3);
    let c12 = nibble(b, 12);
    let c15 = nibble(b, 15);
    if c0 == max || c3 == max || c12 == max || c15 == max {
        (corner_weight * (max as i64)) as f64
    } else {
        0.0
    }
}

/// Σ line(row) + Σ line(col) + corner.
#[inline]
pub fn evaluate(b: u64, table: &[f64], corner_weight: i64) -> f64 {
    let rows = table[(b & 0xFFFF) as usize]
        + table[((b >> 16) & 0xFFFF) as usize]
        + table[((b >> 32) & 0xFFFF) as usize]
        + table[((b >> 48) & 0xFFFF) as usize];
    let cols = table[column(b, 0) as usize]
        + table[column(b, 1) as usize]
        + table[column(b, 2) as usize]
        + table[column(b, 3) as usize];
    let corner = if corner_weight == 0 { 0.0 } else { corner_term(b, corner_weight) };
    rows + cols + corner
}

/// Human-readable decomposition of the evaluation, summed over all 8 lines.
#[derive(Clone, Copy, Debug, Default, PartialEq, Serialize)]
pub struct Breakdown {
    pub empty: i64,
    pub merges: i64,
    pub mono: i64,
    pub sum: i64,
    pub smooth: i64,
    pub stable: i64,
    pub corner: f64,
    pub total: f64,
}

pub fn breakdown(b: u64, table: &[f64], w: &Weights) -> Breakdown {
    let lines: [u16; 8] = [
        (b & 0xFFFF) as u16,
        ((b >> 16) & 0xFFFF) as u16,
        ((b >> 32) & 0xFFFF) as u16,
        ((b >> 48) & 0xFFFF) as u16,
        column(b, 0),
        column(b, 1),
        column(b, 2),
        column(b, 3),
    ];
    let mut acc = Breakdown::default();
    for &v in lines.iter() {
        let f = line_features(v);
        acc.empty += f.empty;
        acc.merges += f.merges;
        acc.mono += f.mono;
        acc.sum += f.sum;
        acc.smooth += f.smooth;
        acc.stable += f.stable;
    }
    acc.corner = corner_term(b, w.corner);
    acc.total = evaluate(b, table, w.corner);
    acc
}
